### ISSUE TITLE
HttpWebRequest sends URI fragment, unlike on .NET Framework

### ISSUE THREAD

**Issue by atifaziz** [dotnet/corefx#39618]
If a `HttpWebRequest` is created with a URL containing a [fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) then the code behaves (i.e. compatibility bug) differently on .NET Core than on .NET Framework.

On .NET Framework, the fragment identifier is clipped and never sent to the server in the HTTP request; On .NET Core, it is. Some web servers, like IIS, reject a URL with a fragment identifier, and so a client that works on .NET Framework can receive a 400 and fail when on targeting .NET Core.

## Bug Demo Program

I have created a self-contained program to demonstrate the issue:

```c#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Runtime.ExceptionServices;
using System.Runtime.InteropServices;

static partial class Program
{
    static int Main(string[] args)
    {
        try
        {
            Console.WriteLine((RuntimeInformation.FrameworkDescription + " ").PadRight(70, '-'));
            return Wain(args);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            return 1;
        }
    }

    // The following client code is identical between .NET Core and .NET
    // Framework versions.

    static class Client
    {
        public static void Run(Uri url)
        {
            var request = WebRequest.CreateHttp(url);

            using var response = (HttpWebResponse) request.GetResponse();
            Console.WriteLine((int)response.StatusCode + " " + response.StatusDescription);

            foreach (var name in response.Headers.AllKeys)
                Console.WriteLine(name + ": " + response.Headers[name]);

            response.GetResponseStream().Close();
        }
    }

    static ExceptionDispatchInfo Try(Action action) =>
        Try<Exception>(action);

    static ExceptionDispatchInfo Try<TError>(Action action) where TError : Exception
    {
        try
        {
            action();
            return null;
        }
        catch (TError e)
        {
            return ExceptionDispatchInfo.Capture(e);
        }
    }

    static int Wain(string[] args)
    {
        if (args.Length == 0)
            throw new Exception("Missing URL argument.");

        var url = new Uri(args[0]);
        var edi = Try(() => Client.Run(url));

#if NETFX // Under .NET Framework, just.

        if (edi != null)
            edi.Throw();

        return 0;

#else // Otherwise, run the .NET Framework version via `dotnet run`

        if (edi != null)
            Console.Error.WriteLine(edi.SourceException);

        // Find the project directory and run the .NET Framework version
        // via `dotnet run`, re-directing standard output and error here.

        var appDir = new DirectoryInfo(AppContext.BaseDirectory);
        var projectDir = appDir.Ascendants().First(dir => dir.EnumerateFiles("*.csproj").Any());

        var psi = new ProcessStartInfo("dotnet", "run --framework net471 " + url)
        {
            CreateNoWindow = true,
            UseShellExecute = false,
            WorkingDirectory = projectDir.FullName,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
        };

        using var process = Process.Start(psi);

        static DataReceivedEventHandler CreateDataReceiverFor(TextWriter writer) => (_, e) =>
        {
            if (e.Data is string line)
                writer.WriteLine(line);
        };

        process.OutputDataReceived += CreateDataReceiverFor(Console.Out);
        process.ErrorDataReceived  += CreateDataReceiverFor(Console.Error);

        process.BeginOutputReadLine();
        process.BeginErrorReadLine();
        process.WaitForExit();

        if (process.ExitCode != 0)
            throw new Exception($"The .NET Framework version of the program exited with a non-zero code of {process.ExitCode}.");

        return edi == null ? 0 : 1;

#endif
    }

    static IEnumerable<DirectoryInfo> Ascendants(this DirectoryInfo dir)
    {
        for (var parent = dir.Parent; parent != null; parent = parent.Parent)
            yield return parent;
    }
}
```

When run as a .NET Core 2.2 application, this program will do the following:

1. It will issue an HTTP request and dump the response status and headers.
2. It will then do the same as 2 but under .NET Framework. This step is done by running the same project via `dotnet run` but with the `--framework net471` option.

I have uploaded a ZIP archive with the full project:

📎 [`UrlFragmentBug.zip`](https://github.com/dotnet/corefx/files/3411028/UrlFragmentBug.zip)

## Steps to Reproduce

1. Decompress [the demo program archive](https://github.com/dotnet/corefx/files/3411028/UrlFragmentBug.zip) and open a Windows Command Prompt or PowerShell session with the current working directory set to the directory of the decompressed project.
2. Start IIS Express using "%ProgramFiles%\IIS Express\iisexpress.exe" /systray:false` on the Windows Command Prompt or `start "$env:ProgramFiles\IIS Express\iisexpress.exe" /systray:false` on the PowerShell prompt.
3. Execute `.\run.cmd http://localhost:8080/`.
4. Execute `.\run.cmd http://localhost:8080/#foo` (notice the addition of the fragment identifier in the URL).

In step 3 you should see two successful HTTP responses for the same URL (`http://localhost:8080/`), under .NET Core and .NET Framework:

    .NET Core 4.6.27817.03 -----------------------------------------------
    200 OK
    Accept-Ranges: bytes
    ETag: "018c519e81cd1:0"
    Server: Microsoft-IIS/10.0
    X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcYXppejFhXE9uZURyaXZlIC0ga29jaGluZC5jb21cRG9jdW1lbnRzXE15IFdlYiBTaXRlc1xXZWJTaXRlMVxpaXNzdGFydC5odG0=?=
    X-Powered-By: ASP.NET
    Date: Fri, 19 Jul 2019 12:27:32 GMT
    Content-Type: text/html
    Last-Modified: Wed, 14 Mar 2012 13:41:04 GMT
    Content-Length: 1307
    .NET Framework 4.7.3416.0 --------------------------------------------
    200 OK
    X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcYXppejFhXE9uZURyaXZlIC0ga29jaGluZC5jb21cRG9jdW1lbnRzXE15IFdlYiBTaXRlc1xXZWJTaXRlMVxpaXNzdGFydC5odG0=?=
    Accept-Ranges: bytes
    Content-Length: 1307
    Content-Type: text/html
    Date: Fri, 19 Jul 2019 12:27:35 GMT
    ETag: "018c519e81cd1:0"
    Last-Modified: Wed, 14 Mar 2012 13:41:04 GMT
    Server: Microsoft-IIS/10.0
    X-Powered-By: ASP.NET

In step 4, with a fragment identifier appended however, you should see a failed HTTP response under .NET Core following by a successful HTTP response under .NET Framework for the _same URL_ (`http://localhost:8080/#foo`):

    .NET Core 4.6.27817.03 -----------------------------------------------
    .\run.cmd : System.Net.WebException: The remote server returned an error: (400) Bad Request.
    At line:1 char:1
    + .\run.cmd http://localhost:8080/#foo 2>&1 |clip
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : NotSpecified: (System.Net.WebE...0) Bad Request.:String) [], RemoteException
        + FullyQualifiedErrorId : NativeCommandError
     
       at System.Net.HttpWebRequest.GetResponse()
       at Program.Client.Run(Uri url) in A:\UrlFragmentBug\Program.cs:line 35
       at Program.<>c__DisplayClass4_0.<Wain>b__0()
     in A:\UrlFragmentBug\Program.cs:line 67
       at Program.Try[TError](Action action) in A:\UrlFragmentBug\Program.cs:line 52

    .NET Framework 4.7.3416.0 --------------------------------------------
    200 OK
    X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcYXppejFhXE9uZURyaXZlIC0ga29jaGluZC5jb21cRG9jdW1lbnRzXE15IFdlYiBTaXRlc1xXZWJTaXRlMVxpaXNzdGFydC5odG0=?=
    Accept-Ranges: bytes
    Content-Length: 1307
    Content-Type: text/html
    Date: Fri, 19 Jul 2019 12:34:37 GMT
    ETag: "018c519e81cd1:0"
    Last-Modified: Wed, 14 Mar 2012 13:41:04 GMT
    Server: Microsoft-IIS/10.0
    X-Powered-By: ASP.NET

## More Information

dotnet --info says:

.NET Core SDK (reflecting any global.json):
 Version:   2.2.204
 Commit:    8757db13ec

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.204\

Host (useful for support):
  Version: 3.0.0-preview7-27902-19
  Commit:  fbe9466ddd

.NET Core SDKs installed:
  1.1.13 [C:\Program Files\dotnet\sdk]
  1.1.14 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.502 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.1.507 [C:\Program Files\dotnet\sdk]
  2.1.508 [C:\Program Files\dotnet\sdk]
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.1.604 [C:\Program Files\dotnet\sdk]
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.701 [C:\Program Files\dotnet\sdk]
  2.2.101 [C:\Program Files\dotnet\sdk]
  2.2.202 [C:\Program Files\dotnet\sdk]
  2.2.204 [C:\Program Files\dotnet\sdk]
  2.2.300 [C:\Program Files\dotnet\sdk]
  2.2.301 [C:\Program Files\dotnet\sdk]
  3.0.100-preview7-012802 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview7.19353.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview7-27902-19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview7-27902-19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by stephentoub** [comment]
Thanks.  Can you try on .NET Core 3.0?  I expect this was fixed by https://github.com/dotnet/corefx/pull/32574.

**Comment by atifaziz** [comment]
@stephentoub After making the following changes to [the attached demo](https://github.com/dotnet/corefx/files/3411028/UrlFragmentBug.zip):

```diff
diff --git a/BugDemo.csproj b/BugDemo.csproj
index feb1f31..732d0af 100644
--- a/BugDemo.csproj
+++ b/BugDemo.csproj
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.2;net471</TargetFrameworks>
+    <TargetFrameworks>netcoreapp3.0;net471</TargetFrameworks>
     <OutputType>Exe</OutputType>
     <LangVersion>8</LangVersion>
   </PropertyGroup>
diff --git a/global.json b/global.json
index 7249be5..86298ca 100644
--- a/global.json
+++ b/global.json
@@ -1,5 +1,5 @@
 {
   "sdk": {
-    "version": "2.2.200"
+    "version": "3.0.100-preview7-012802"
   }
 }
\ No newline at end of file
diff --git a/run.cmd b/run.cmd
index 92b9613..e58ce78 100644
--- a/run.cmd
+++ b/run.cmd
@@ -1,4 +1,4 @@
 @echo off
 pushd "%~dp0"
-dotnet run --framework netcoreapp2.2 -- %*
+dotnet run --framework netcoreapp3.0 -- %*
 popd
```

Then running with `.\run.cmd http://localhost:8080/#foo` produced the following output:

```
.NET Core 3.0.0-preview7-27902-19 ------------------------------------
200 OK
Accept-Ranges: bytes
ETag: "018c519e81cd1:0"
Server: Microsoft-IIS/10.0
X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcYXppejFhXE9uZURyaXZlIC0ga29jaGluZC5jb21cRG9jdW1lbnRzXE15IFdlYiBTaXRlc1xXZWJTaXRlMVxpaXNzdGFydC5odG0=?=
X-Powered-By: ASP.NET
Date: Fri, 19 Jul 2019 15:11:42 GMT
Content-Type: text/html
Last-Modified: Wed, 14 Mar 2012 13:41:04 GMT
Content-Length: 1307
.NET Framework 4.7.3416.0 --------------------------------------------
200 OK
X-SourceFiles: =?UTF-8?B?QzpcVXNlcnNcYXppejFhXE9uZURyaXZlIC0ga29jaGluZC5jb21cRG9jdW1lbnRzXE15IFdlYiBTaXRlc1xXZWJTaXRlMVxpaXNzdGFydC5odG0=?=
Accept-Ranges: bytes
Content-Length: 1307
Content-Type: text/html
Date: Fri, 19 Jul 2019 15:11:44 GMT
ETag: "018c519e81cd1:0"
Last-Modified: Wed, 14 Mar 2012 13:41:04 GMT
Server: Microsoft-IIS/10.0
X-Powered-By: ASP.NET
```

The fragment identifier doesn't cause an error so I can confirm that the bug fix seems to be in .NET Core 3.0. Does this mean it will eventually find its way into a .NET Core 2.x update/patch?

**Comment by stephentoub** [comment]
> Does this mean it will eventually find its way into a .NET Core 2.x update/patch?

Probably not, but @karelz can comment.

>  so I can confirm that the bug fix that is #32574 seems to be in .NET Core 3.0

Glad it addressed the issue for you.  Thanks for confirming.
